### ISSUE TITLE
x/vuln: decentralized option

### ISSUE THREAD

**Issue by willfaught** [golang/go#54982]
(This is a suggestion, which the Go blog post about vuln asked for and linked to here, so I'm ignoring the issue template.)

While I like that vuln reports are tailored and cleaned up by the Go Team, I think the community would be best served with an option for a decentralized solution, at least as a fallback, similar to how the module proxy and sum database work.

Module authors should be able to publish their own vuln determinations as part of the module or package itself, and those publications should be able to be extracted as part of the module/package info. That way the community isn't reliant on the Go Team to communicate and manage vulns.

**Comment by seankhliao** [comment]
This is already in place, see [package docs](https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck):

> Set the GOVULNDB environment variable to specify a different database, which must implement the specification at https://go.dev/security/vuln/database.

**Comment by seankhliao** [comment]
The distributed mechanism is to publish retractions.

Vulnerability reporting relies on somewhat centralized CVE IDs to allow communication between different parties, a fully decentralized option doesn't make sense here.

**Comment by seankhliao** [comment]
cc @golang/vulndb

**Comment by zpavlinovic** [comment]
As mentioned earlier, users can provide other vulnerability databases to `govulncheck` via `GOVULNDB`  environment variable.

**Comment by willfaught** [comment]
I wasn't clear, I think. I don't mean in the sense of GOPROXY=myproxy, I mean in the sense of GOPROXY=direct. Basically, have a way for module authors to declare vulns in modules themselves, identified by commit hash or version, e.g.

```
module example.com/foo

go 1.19

vulnerability {
  versions: v1.2.0...v1.2.9, v1.2.11
  description: "My description"
  method: Bar.Baz
}
```

**Comment by willfaught** [comment]
Common identifiers can be derived from the module and vulnerability info, e.g. the `example.com/foo:Bar.Baz:v1.2.0...v1.2.9,v1.2.11` vulnerability. Or maybe there's a naming convention like

```
module example.com/foo

go 1.19

vulnerability 2022-09-09 {
  versions: v1.2.0...v1.2.9, v1.2.11
  description: "My description"
  method: Bar.Baz
}
```

Vulnerability `example.com/foo:2022-09-09`.

**Comment by seankhliao** [comment]
That was what eventually became the `retract` directive: https://github.com/golang/go/issues/24031

**Comment by zpavlinovic** [comment]
I believe this would need to go through a [formal proposal process](https://github.com/golang/proposal) with pros, cons, motivation, etc.

**Comment by willfaught** [comment]
>That was what eventually became the retract directive

Thanks for pointing me to that.

- It looks like [the retract directive is ignored by Go 1.15 and earlier](https://go.dev/ref/mod#go-mod-file-retract):

    >Go 1.15 and lower will report an error if a retract directive is written in the [main module’s](https://go.dev/ref/mod#glos-main-module) go.mod file and will ignore retract directives in go.mod files of dependencies.

    Wouldn't that be an issue when using the vuln tool for those Go versions?

- Retractions don't seem to be necessarily considered a security concern by the vuln tool/database, and aren't considered unless you upgrade the package manually or by Minimum Version Selection to a version that has the retraction
- Retractions don't flag particular functions/methods, as the vuln tool/database does

>believe this would need to go through a [formal proposal process](https://github.com/golang/proposal) with pros, cons, motivation, etc.

If this is true, then we should fix the Go blog post to not point users to make vuln suggestions here by creating an `x/vuln:` issue. This isn't meant to be a suggestion in the formal Go proposal process sense.

**Comment by zpavlinovic** [comment]
> If this is true, then we should fix the Go blog post to not point users to make vuln suggestions here by creating an `x/vuln:` issue. This isn't meant to be a suggestion in the formal Go proposal process sense.

I should have been more precise. My mention of the formal proposal process was in response to the suggested change
```
module example.com/foo

go 1.19

vulnerability 2022-09-09 {
  versions: v1.2.0...v1.2.9, v1.2.11
  description: "My description"
  method: Bar.Baz
}
```
which would also require modifications to the specification and handling of `go.mod` files. However, we can choose a different way to implement this.

My concern is that this would add an extra maintenance burden on the programmers so this would not completely replace the existing approach. Or maybe it is not supposed to?

**Comment by willfaught** [comment]
>My concern is that this would add an extra maintenance burden on the programmers

I think this would actually be simpler for module owners. Instead of figuring out how to report a CVE, a GHABCDEFwhateverthisis, submit a report to the official Go vuln database, etc. etc., they can simply report the problem in go.mod, and the toolchain automatically surfaces the problem when others try to build with it. All the info is contained within the module itself, and published as one unit.

I guess the go tool would have to consult the latest version for vuln retractions regardless of what the minimum version selected for builds is.

>so this would ultimately not be able to completely replace the existing approach

Which approach do you mean? The new vuln stuff?

**Comment by zpavlinovic** [comment]
> > My concern is that this would add an extra maintenance burden on the programmers
> 
> I think this would actually be simpler for module owners. Instead of figuring out how to report a CVE, a GHABCDEFwhateverthisis, submit a report to the official Go vuln database, etc. etc., they can simply report the problem in go.mod, and the toolchain automatically surfaces the problem when others try to build with it. All the info is contained within the module itself, and published as one unit.
> 
> I guess the go tool would have to consult the latest version for vuln retractions regardless of what the minimum version selected for builds is.
> 
> > so this would ultimately not be able to completely replace the existing approach
> 
> Which approach do you mean? The new vuln stuff?

Yes, the current approach with a vulnerability database.

@tatianab @jba @neild

**Comment by willfaught** [comment]
I would say that a decentralized approach would be a good basis, but you'd also want users to be able to specify a database they trust (defaulting to the Go Team's) for issuing retractions on behalf of module owners, like if module owners are too slow to do it.

Perhaps if vuln retractions are cached in go.sum or something comparable to prevent bad actors from undoing module-based retractions, then database entries could be removed as the corresponding modules publish retractions. This keeps the "active" database size to a minimum, and keeps the vuln retraction history with the (cached) modules.

**Comment by mwriter** [comment]
> via `GOVULNDB` environment variable

Is it still relevant?

**Comment by seankhliao** [comment]
I don't think there's anything to do here.

**Comment by mwriter** [comment]
> I don't think there's anything to do here.

It looks like you've become a fan of mine. It's inspiring.

**Comment by rolandshoemaker** [comment]
> Is it still relevant?

Yes, you can still use GOVULNDB to specify your own database, as long as it conforms to the format specified in https://go.dev/doc/security/vuln/database.

**Comment by mwriter** [comment]
> Yes, you can still use GOVULNDB to specify your own database, as long as it conforms to the format specified in https://go.dev/doc/security/vuln/database.

Hi. I just tried it and it didn't work. In addition, I don't see any mention `GOVULNDB` in the code `golang.org/x/vuln`.

**Comment by rolandshoemaker** [comment]
Oh hm, you are correct. At some point we must've dropped that. `govulncheck` has a `-db` flag that operates in the same fashion.

**Comment by mwriter** [comment]
> At some point we must've dropped that. `govulncheck` has a `-db` flag that operates in the same fashion.

That's what I'd like to talk about in detail. Do you mind continuing?

**Comment by rolandshoemaker** [comment]
What would you like to know?

**Comment by mwriter** [comment]
> What would you like to know?

First of all, I would like to discuss the possibility of returning the setting of the environment variable. Golang is based on flexible work with variable environments. And it's strange that in this case such a refactoring option is chosen.

**Comment by rolandshoemaker** [comment]
That seems reasonable. I've opened https://go.dev/issue/77080 for that.

**Comment by mwriter** [comment]
> That seems reasonable. I've opened https://go.dev/issue/77080 for that.

Wow, thanks!
I thought about opening it myself, but your "leader" would probably have run to close it right away :)

**Comment by mwriter** [comment]
Then one more question on this topic. What do you think about the possibility of adding support relative db paths too?
